pkg/discovery/systemd.go

```diff
@@ -50,111 +45,66 @@ func (c *SystemdConfig) NewDiscoverer(d DiscovererOptions) (Discoverer, error) {
 }
 
 func (c *SystemdDiscoverer) Run(ctx context.Context, up chan<- []*Group) error {
-	ticker := time.NewTicker(time.Second)
-	defer ticker.Stop()
-
-	for {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case <-ticker.C:
-		}
-
-		units, err := c.units(ctx)
-		if err != nil {
-			return fmt.Errorf("failed to list units: %w", err)
-		}
-		newSourceList := map[string]struct{}{}
-		var targetGroups []*Group
-		for _, unit := range units {
-			labelset := model.LabelSet{
-				"systemd_unit": model.LabelValue(unit),
-			}
-			pids, err := c.pids(ctx, unit)
-			if err != nil {
-				level.Debug(c.logger).Log("msg", "failed to get PIDs for unit", "unit", unit, "err", err)
-				continue
-			}
-
-			newSourceList[unit] = struct{}{}
-
-			targetGroups = append(targetGroups, &Group{
-				Targets: []model.LabelSet{{}},
-				Labels:  labelset,
-				Source:  unit,
-				PIDs:    pids,
-			})
-		}
-
-		// Add empty groups for targets which have been removed since the previous run.
-		for unit := range c.oldSourceList {
-			if _, ok := newSourceList[unit]; !ok {
-				targetGroups = append(targetGroups, &Group{Source: unit})
-			}
-		}
-		c.oldSourceList = newSourceList
-
-		up <- targetGroups
-	}
-}
-
-func (c *SystemdDiscoverer) units(ctx context.Context) ([]string, error) {
-	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
-	defer cancel()
-
-	cmd := exec.CommandContext(ctx, "systemctl", "list-units", "--type=service", "--state=running")
-	stdout, err := cmd.StdoutPipe()
+	conn, err := systemd.NewWithContext(ctx)
 	if err != nil {
-		return nil, err
-	}
-	if err := cmd.Start(); err != nil {
-		return nil, err
+		level.Error(c.logger).Log("msg", "failed to connect to systemd D-Bus API", "err", err)
 	}
+	defer conn.Close()
 
-	units := parseUnitList(stdout)
-	if err := cmd.Wait(); err != nil {
-		return nil, err
+	isSubStateChanged := func(u1, u2 *systemd.UnitStatus) bool {
+		return u1.SubState != u2.SubState
 	}
-	return units, nil
-}
 
-func parseUnitList(stdout io.Reader) []string {
-	units := []string{}
-	s := bufio.NewScanner(stdout)
-	for s.Scan() {
-		line := strings.TrimSpace(s.Text())
-		if strings.HasPrefix(line, "UNIT") ||
-			strings.HasPrefix(line, "LOAD") ||
-			strings.HasPrefix(line, "ACTIVE") ||
-			strings.HasPrefix(line, "SUB") {
-			continue
-		}
-		fields := strings.Fields(line)
-		if len(fields) < 1 {
-			continue
-		}
-		units = append(units, fields[0])
+	isService := func(name string) bool {
+		return !strings.HasSuffix(name, ".service")
```

Comment:
```suggestion
		return !strings.HasSuffix(name, ".service")
```
Shouldn't it be like this, as the name suggests?

Reply:
I think you forgot to suggest a change :grin:

I know this works as intended, do you mean the function should be called `isNotService`?
I meant "I want the services only", but I was surprised that it needs a negative filter function

Reply:
Yeah, exactly that (I've also fixed the suggestion). Semantically name and what it does not resonate for me 😅